data/appRateLimiter.go

```diff
@@ -0,0 +1,249 @@
+// in order to make it testable
+func (r *appRateLimiter) shouldDropInner(txgroup []transactions.SignedTxn, origin []byte, now time.Time) bool {
+	buckets, keys := txgroupToKeys(txgroup, origin, r.seed, r.salt, numBuckets)
+	if len(keys) == 0 {
+		return false
+	}
+	return r.shouldDropKeys(buckets, keys, now)
+}
+
+func (r *appRateLimiter) shouldDropKeys(buckets []int, keys []keyType, now time.Time) bool {
+	curInt := r.interval(now)
+
+	for i := range keys {
+		key := keys[i]
+		bucket := buckets[i]
+		entry, has := r.entry(bucket, key, curInt)
+		if !has {
+			// new entry, defaults are provided by entry() function
+			continue
+		}
+
+		interval := entry.interval.Load()
+		if interval != curInt {
+			var val uint64 = 0
+			if interval == curInt-1 {
+				// there are continuous intervals, use the previous value
+				val = entry.cur.Load()
+			}
+			entry.prev.Store(val)
+			entry.cur.Store(1)
+			entry.interval.Store(curInt)
+		} else {
+			entry.cur.Add(1)
+		}
+
+		curFraction := r.fraction(now)
+		rate := uint64(float64(entry.prev.Load())*(1-curFraction)) + entry.cur.Load()
+
+		if rate > r.serviceRatePerWindow {
+			return true
+		}
+	}
+
+	return false
+}
+
+// txgroupToKeys converts txgroup data to keys
+func txgroupToKeys(txgroup []transactions.SignedTxn, origin []byte, seed uint64, salt [16]byte, numBuckets int) ([]int, []keyType) {
+	// there are max 16 * 8 = 128 apps (buckets, keys) per txgroup
+	// TODO: consider sync.Pool
+
+	var keys []keyType
+	var buckets []int
+	// since blake2 is a crypto hash function it seems OK to shrink 32 bytes digest down to 8.
+	// Rationale: we expect thousands of apps sent from thousands of peers,
+	// so required millions of unique pairs => 8 bytes should be enough.
+	// The 16 bytes salt makes it harder to find collisions if an adversary attempts to censor
+	// some app by finding a collision with some app and flood a network with such transactions:
+	// h(app + relay_ip) = h(app2 + relay_ip).
+	var buf [8 + 16 + 16]byte // uint64 + 16 bytes of salt + up to 16 bytes of address
+	txnToDigest := func(appIdx basics.AppIndex) keyType {
+		binary.LittleEndian.PutUint64(buf[:8], uint64(appIdx))
+		copy(buf[8:], salt[:])
+		copied := copy(buf[8+16:], origin)
+
+		h := blake2b.Sum256(buf[:8+16+copied])
+		var key keyType
+		copy(key[:], h[:len(key)])
+		return key
+	}
+	for i := range txgroup {
+		if txgroup[i].Txn.Type == protocol.ApplicationCallTx {
+			appIdx := txgroup[i].Txn.ApplicationID
+			// hash appIdx into a bucket, do not use modulo since it could
+			// assign two vanilla (and presumable, popular) apps to the same bucket.
+			buckets = append(buckets, int(memhash64(uint64(appIdx), seed)%uint64(numBuckets)))
+			keys = append(keys, txnToDigest(appIdx))
+			if len(txgroup[i].Txn.ForeignApps) > 0 {
+				for _, appIdx := range txgroup[i].Txn.ForeignApps {
+					buckets = append(buckets, int(memhash64(uint64(appIdx), seed)%uint64(numBuckets)))
+					keys = append(keys, txnToDigest(appIdx))
+				}
```

Comment:
maybe since you're doing lambdas anyway maybe you add 
```
txnToBucket := func(appIdx basics.AppIndex) int {
     return int(memhash64(uint64(appIdx), seed)%uint64(numBuckets)
}
```
then you can avoid the double-definition below

Reply:
done